blosc/schunk.c

```diff
@@ -293,41 +279,108 @@ blosc2_schunk* blosc2_schunk_copy(blosc2_schunk *schunk, blosc2_storage storage)
       BLOSC_TRACE_ERROR("Can not update the `usermeta`.");
       return NULL;
     }
+    free(usermeta);
   }
   return new_schunk;
 }
 
 
 /* Open an existing super-chunk that is on-disk (no copy is made). */
-blosc2_schunk* blosc2_schunk_open(const blosc2_storage storage) {
-  if (storage.urlpath == NULL) {
-    BLOSC_TRACE_ERROR("You need to supply a storage.urlpath.");
+blosc2_schunk* blosc2_schunk_open(char* urlpath) {
+  if (urlpath == NULL) {
+    BLOSC_TRACE_ERROR("You need to supply a urlpath.");
     return NULL;
   }
 
-  // We only support frames yet
-  blosc2_frame* frame = blosc2_frame_from_file(storage.urlpath);
-  blosc2_schunk* schunk = blosc2_frame_to_schunk(frame, false);
+  blosc2_frame_s* frame = frame_from_file(urlpath);
+  blosc2_schunk* schunk = frame_to_schunk(frame, false);
 
-  // Get the storage with proper defaults
-  blosc2_cparams *store_cparams;
-  blosc2_schunk_get_cparams(schunk, &store_cparams);
-  blosc2_dparams *store_dparams;
-  blosc2_schunk_get_dparams(schunk, &store_dparams);
-  schunk->storage = get_new_storage(&storage, store_cparams, store_dparams);
-  free(store_cparams);
-  free(store_dparams);
+  // Set the storage with proper defaults
+  size_t pathlen = strlen(urlpath);
+  schunk->storage->urlpath = malloc(pathlen + 1);
+  strcpy(schunk->storage->urlpath, urlpath);
+  schunk->storage->sequential = !frame->sframe;
   // Update the existing cparams/dparams with the new defaults
   update_schunk_properties(schunk);
 
   return schunk;
 }
 
 
+int64_t blosc2_schunk_to_buffer(blosc2_schunk* schunk, uint8_t** dest, bool* needs_free) {
+  blosc2_frame_s* frame;
+  int64_t framebuf_len;
+  if ((schunk->storage->sequential == true) && (schunk->storage->urlpath == NULL)) {
+    frame =  (blosc2_frame_s*)(schunk->frame);
+    *dest = frame->framebuf;
+    framebuf_len = frame->len;
+    *needs_free = false;
+  }
+  else {
+    // Copy to a sequential storage
+    blosc2_storage frame_storage = {.sequential=true};
+    blosc2_schunk* schunk_copy = blosc2_schunk_copy(schunk, frame_storage);
+    if (schunk_copy == NULL) {
+      BLOSC_TRACE_ERROR("Error during the conversion of schunk to buffer.");
+      return BLOSC2_ERROR_SCHUNK_COPY;
+    }
+    frame = (blosc2_frame_s*)(schunk_copy->frame);
+    *dest = frame->framebuf;
+    framebuf_len = frame->len;
+    *needs_free = true;
+    frame->avoid_framebuf_free = true;
+    blosc2_schunk_free(schunk_copy);
+  }
+
+  return framebuf_len;
+
+}
+
+
+/* Write an in-memory frame out to a file. */
+int64_t frame_to_file(blosc2_frame_s* frame, char* urlpath) {
+  FILE* fp = fopen(urlpath, "wb");
+  size_t nitems = fwrite(frame->framebuf, (size_t)frame->len, 1, fp);
+  fclose(fp);
+  return nitems * (size_t)frame->len;
+}
+
+
+/* Write super-chunk out to a file. */
+int64_t blosc2_schunk_to_file(blosc2_schunk* schunk, char* urlpath) {
+  if (urlpath == NULL) {
+    BLOSC_TRACE_ERROR("urlpath cannot be NULL");
+    return -1;
```

Comment:
Use `BLOSC2_ERROR_INVALID_PARAM` instead of -1.

Reply:
Fixed

Reply:
Fixed.